src/clj/witan/send/params.clj

```diff
@@ -44,11 +44,16 @@
           params (sort academic-years)))
 
 (defn calculate-joiners-per-calendar-year
+  "The result maps keys are used to lookup up external population so need to be added
+   even though there is no observed count"
   [transitions]
+  (let [observed-calendar-years (into #{} (map :calendar-year transitions))
+        joiners-per-calendar-year (into {} (for [cy observed-calendar-years] [cy {}]))]
```

Comment:
I'm not sure what this bit is doing?

Reply:
It creates a set of observed calendar years from all the transitions data and then creates an empty set of results but with keys of each calendar year i.e, `{2016: {}, 2017: {}, 2018: {}}`

The results now look of the function now look something like `{2016: {1: 20, 2:34,..}, 2017: {}, 2018: {1: 20, 2:34,..}}` rather than `{2016: {1: 20, 2:34,..}, 2018: {1: 20, 2:34,..}}`.  Having this key present is enough for the later function that looks up the non-joiners for 2017 to work properly.